src/main/java/de/numcodex/feasibility_gui_backend/query/broker/direct/DirectBrokerClientCql.java

```diff
@@ -0,0 +1,186 @@
+import org.hl7.fhir.r4.model.Library;
+import org.hl7.fhir.r4.model.Measure;
+import org.hl7.fhir.r4.model.MeasureReport;
+import org.hl7.fhir.r4.model.Parameters;
+
+import java.io.IOException;
+import java.util.*;
+
+import static de.numcodex.feasibility_gui_backend.query.QueryMediaType.CQL;
+import static de.numcodex.feasibility_gui_backend.query.collect.QueryStatus.COMPLETED;
+import static de.numcodex.feasibility_gui_backend.query.collect.QueryStatus.FAILED;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.hl7.fhir.r4.model.Bundle.BundleType.TRANSACTION;
+import static org.hl7.fhir.r4.model.Bundle.HTTPVerb.POST;
+
+/**
+ * A {@link BrokerClient} to be used to directly communicate with a CQL-capable FHIR Server instance
+ * without the need for using any middleware (Aktin or DSF).
+ */
+@Slf4j
+public class DirectBrokerClientCql extends DirectBrokerClient {
+
+    private static final String SITE_1_NAME = "CQL Server";
+    private final FhirContext fhirContext;
+    private final IGenericClient fhirClient;
+
+    /**
+     * Creates a new {@link DirectBrokerClientCql} instance that uses the given web client to
+     * communicate with a CQL capable FHIR server instance.
+     *
+     * @param fhirContext A FHIR context.
+     * @param fhirClient A FHIR client, configured with the correct url
+     */
+    public DirectBrokerClientCql(FhirContext fhirContext, IGenericClient fhirClient) {
+        this.fhirContext = Objects.requireNonNull(fhirContext);
+        this.fhirClient = Objects.requireNonNull(fhirClient);
+        listeners = new ArrayList<>();
+        brokerQueries = new HashMap<>();
+        brokerToBackendQueryIdMapping = new HashMap<>();
+    }
+
+    @Override
+    public void publishQuery(String brokerQueryId) throws QueryNotFoundException, IOException {
+        var query = findQuery(brokerQueryId);
+        var queryContent = Optional.ofNullable(query.getQueryDefinition(CQL))
+            .orElseThrow(() -> new IllegalStateException("Query with ID "
+                + query.getQueryId()
+                + " does not contain a query definition for the mandatory type: "
+                + CQL
+            ));
+        try {
+            updateQueryStatus(query, QueryStatus.EXECUTING);
+            var measureUri = createMeasureAndLibrary(queryContent);
+            var report = evaluateMeasure(measureUri);
+            var resultCount = report.getGroupFirstRep().getPopulationFirstRep().getCount();
+            query.registerSiteResults(SITE_1_ID, obfuscateResultCount ? obfuscate(resultCount) : resultCount);
+            updateQueryStatus(query, COMPLETED);
+        } catch (Exception e) {
+            updateQueryStatus(query, FAILED);
+            throw new IOException("An error occurred while publishing the query with ID: " + query.getQueryId(), e);
+        }
+    }
+
+    @Override
+    public String getSiteName(String siteId) {
+        return siteId.equals(SITE_1_ID) ? SITE_1_NAME : "";
```

Comment:
Shouldn't `getSiteName` throw an `SiteNotFoundException` instead of returning an empty string?

Reply:
That implementation can be moved into `DirectBrokerClient`. Also reversing equals saves us from an NPE here.